File: app/src/main/java/com/mvvm/contactlist/utilities/InjectorUtils.java
package com.mvvm.contactlist.utilities;

import android.content.Context;

import com.mvvm.contactlist.db.AppDatabase;
import com.mvvm.contactlist.db.dao.CategoryDao;
import com.mvvm.contactlist.db.dao.ContactDao;
import com.mvvm.contactlist.repo.CategoryRepository;
import com.mvvm.contactlist.repo.ContactRepository;
import com.mvvm.contactlist.viewmodel.CategoryViewModel;
import com.mvvm.contactlist.viewmodel.ContactListViewModel;
import com.mvvm.contactlist.viewmodel.ContactViewModel;

public class InjectorUtils {

    public static CategoryViewModel.AddCategoryViewModelFactory provideAddCategoryViewModelFactory(Context context) {
        CategoryDao dao = AppDatabase.getDatabase(context.getApplicationContext()).categoryDao();
        CategoryRepository repository = CategoryRepository.getInstance(dao);
        return new CategoryViewModel.AddCategoryViewModelFactory(repository);
    }

    public static ContactViewModel.ContactViewModelFactory provideContactViewModelFactory(Context context) {
        CategoryDao categoryDao = AppDatabase.getDatabase(context.getApplicationContext()).categoryDao();
        ContactDao contactDao = AppDatabase.getDatabase(context.getApplicationContext()).contactDao();
        ContactRepository repository = ContactRepository.getInstance(categoryDao, contactDao);
        return new ContactViewModel.ContactViewModelFactory(repository);
    }

    public static ContactListViewModel.ContactListViewModelFactory provideContactListViewModelFactory(Context context) {
        CategoryDao categoryDao = AppDatabase.getDatabase(context.getApplicationContext()).categoryDao();
        ContactDao contactDao = AppDatabase.getDatabase(context.getApplicationContext()).contactDao();
        ContactRepository repository = ContactRepository.getInstance(categoryDao, contactDao);
        return new ContactListViewModel.ContactListViewModelFactory(repository);
    }
}

File: app/src/main/java/com/mvvm/contactlist/utilities/InjectorUtils.kt
package com.mvvm.contactlist.utilities

import android.content.Context
import com.mvvm.contactlist.db.AppDatabase.Companion.getDatabase
import com.mvvm.contactlist.repo.CategoryRepository
import com.mvvm.contactlist.repo.ContactRepository
import com.mvvm.contactlist.viewmodel.CategoryViewModel.AddCategoryViewModelFactory
import com.mvvm.contactlist.viewmodel.ContactListViewModel.ContactListViewModelFactory
import com.mvvm.contactlist.viewmodel.ContactViewModel.ContactViewModelFactory

object InjectorUtils {
    fun provideAddCategoryViewModelFactory(context: Context): AddCategoryViewModelFactory {
        val dao = getDatabase(context.applicationContext).categoryDao()
        val repository = CategoryRepository.getInstance(dao)
        return AddCategoryViewModelFactory(repository)
    }

    fun provideContactViewModelFactory(context: Context): ContactViewModelFactory {
        val categoryDao = getDatabase(context.applicationContext).categoryDao()
        val contactDao = getDatabase(context.applicationContext).contactDao()
        val repository = ContactRepository.getInstance(categoryDao, contactDao)
        return ContactViewModelFactory(repository)
    }

    fun provideContactListViewModelFactory(context: Context): ContactListViewModelFactory {
        val categoryDao = getDatabase(context.applicationContext).categoryDao()
        val contactDao = getDatabase(context.applicationContext).contactDao()
        val repository = ContactRepository.getInstance(categoryDao, contactDao)
        return ContactListViewModelFactory(repository)
    }
}